File: packages/report-export-engine/src/infrastructure/pdf/latex-compiler.ts
import { copyFileSync, existsSync, mkdtempSync, readFileSync, rmSync, writeFileSync, mkdirSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";

import { resolveLatexAssetRegistry, type LatexAssetRegistry } from "../assets/asset-registry";

export type LatexCompilationErrorKind = "compiler_missing" | "compilation_failed" | "output_missing";

export class LatexCompilationError extends Error {
  kind: LatexCompilationErrorKind;
  logs: string;
  workDir?: string;

  constructor(
    kind: LatexCompilationErrorKind,
    message: string,
    logs = "",
    workDir?: string,
  ) {
    super(message);
    this.name = "LatexCompilationError";
    this.kind = kind;
    this.logs = logs;
    this.workDir = workDir;
  }
}

export interface CompileLatexToPdfOptions {
  texContent: string;
  timeoutMs?: number;
  jobName?: string;
  cleanup?: boolean;
  preserveWorkDirOnError?: boolean;
  assets?: LatexAssetRegistry;
  extraFiles?: Array<{
    relativePath: string;
    content: string | Buffer;
  }>;
}

export interface CompileLatexToPdfResult {
  pdfBuffer: Buffer;
  logs: string;
}

export function isPdflatexAvailable(): boolean {
  const command = spawnSync("pdflatex", ["--version"], {
    encoding: "utf8",
  });
  return command.status === 0;
}

function runPdflatexPass(
  workDir: string,
  texFilePath: string,
  timeoutMs: number,
): { status: number | null; output: string } {
  const run = spawnSync(
    "pdflatex",
    [
      "-interaction=nonstopmode",
      "-halt-on-error",
      "-file-line-error",
      path.basename(texFilePath),
    ],
    {
      cwd: workDir,
      encoding: "utf8",
      timeout: timeoutMs,
    },
  );

  const output = `${run.stdout || ""}\n${run.stderr || ""}`;
  return { status: run.status, output };
}

export function compileLatexToPdf(options: CompileLatexToPdfOptions): CompileLatexToPdfResult {
  const timeoutMs = options.timeoutMs ?? 120_000;
  const jobName = options.jobName ?? "report";
  const cleanup = options.cleanup ?? true;

  if (!isPdflatexAvailable()) {
    throw new LatexCompilationError(
      "compiler_missing",
      "pdflatex is not available in the current environment.",
    );
  }

  const assets = options.assets ?? resolveLatexAssetRegistry();
  const workDir = mkdtempSync(path.join(tmpdir(), "aalie-export-"));
  const texFilePath = path.join(workDir, `${jobName}.tex`);
  const logosOutputDir = path.join(workDir, "logos");

  const logs: string[] = [];
  let shouldCleanup = cleanup;

  try {
    mkdirSync(logosOutputDir, { recursive: true });

    copyFileSync(assets.styleFilePath, path.join(workDir, "aalie-report.sty"));
    copyFileSync(assets.ucaldasLogoPath, path.join(logosOutputDir, "ucaldas.pdf"));
    copyFileSync(assets.aalieLogoPath, path.join(logosOutputDir, "aalie.pdf"));

    for (const file of options.extraFiles || []) {
      const rel = String(file.relativePath || "").trim().replace(/^\/+/, "");
      if (!rel) continue;
      const dest = path.join(workDir, rel);
      mkdirSync(path.dirname(dest), { recursive: true });
      writeFileSync(dest, file.content);
    }

    writeFileSync(texFilePath, options.texContent, "utf8");

    for (let pass = 1; pass <= 2; pass += 1) {
      const run = runPdflatexPass(workDir, texFilePath, timeoutMs);
      logs.push(`--- pdflatex pass ${pass} ---\n${run.output}`);

      if (run.status !== 0) {
        throw new LatexCompilationError(
          "compilation_failed",
          `pdflatex failed on pass ${pass} with status ${String(run.status)}.`,
          logs.join("\n"),
        );
      }
    }

    const pdfPath = path.join(workDir, `${jobName}.pdf`);
    if (!existsSync(pdfPath)) {
      throw new LatexCompilationError(
        "output_missing",
        `Expected PDF output was not generated at ${pdfPath}.`,
        logs.join("\n"),
      );
    }

    const pdfBuffer = readFileSync(pdfPath);
    return {
      pdfBuffer,
      logs: logs.join("\n"),
    };
  } catch (error) {
    if (error instanceof LatexCompilationError) {
      if (options.preserveWorkDirOnError) {
        error.workDir = workDir;
        shouldCleanup = false;
      }
      throw error;
    }
    throw error;
  } finally {
    if (shouldCleanup) {
      rmSync(workDir, { recursive: true, force: true });
    }
  }
}

File: apps/api/app/modules/export/latex_compiler.py
"""
PDF compilation via pdflatex with rich diagnostics.
"""

from __future__ import annotations

import shutil
import subprocess
import tempfile
import time
from pathlib import Path
from typing import Dict, Iterable, List

from .asset_builder import build_asset_manifest
from .asset_registry import resolve_latex_asset_registry
from .models import LatexCompilationError


def is_pdflatex_available() -> bool:
    return shutil.which("pdflatex") is not None


def compile_latex_to_pdf(
    tex_content: str,
    *,
    timeout_ms: int | None = None,
    job_name: str | None = None,
    cleanup: bool = True,
    extra_files: Iterable[Dict[str, bytes | str]] | None = None,
    preserve_workdir_on_error: bool = False,
    passes: int = 2,
) -> Dict[str, object]:
    timeout = (timeout_ms or 120_000) / 1000.0
    name = job_name or "report"
    if passes not in {1, 2}:
        raise ValueError("passes must be 1 or 2")

    if not is_pdflatex_available():
        raise LatexCompilationError(
            "compiler_missing",
            "pdflatex is not available in the current environment.",
        )

    assets = resolve_latex_asset_registry()
    profile: Dict[str, float] = {}
    setup_started = time.perf_counter()
    work_dir = Path(tempfile.mkdtemp(prefix="aalie-export-"))
    profile["temp_setup_ms"] = (time.perf_counter() - setup_started) * 1000
    tex_path = work_dir / f"{name}.tex"
    pdf_path = work_dir / f"{name}.pdf"
    logos_dir = work_dir / "logos"
    logs: List[str] = []
    asset_entries: List[Dict[str, object]] = []

    def _copy_file(src: str, dest: Path) -> None:
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copyfile(src, dest)
        asset_entries.append(
            {
                "filename": str(dest.relative_to(work_dir)),
                "mimeType": ("application/pdf" if dest.suffix.lower() == ".pdf" else "text/plain"),
                "size": dest.stat().st_size,
            }
        )

    try:
        logos_dir.mkdir(parents=True, exist_ok=True)
        _copy_file(assets.style_file_path, work_dir / "aalie-report.sty")
        _copy_file(assets.ucaldas_logo_path, logos_dir / "ucaldas.pdf")
        _copy_file(assets.aalie_logo_path, logos_dir / "aalie.pdf")

        for extra in extra_files or []:
            rel = str(extra.get("relativePath") or "").lstrip("/").strip()
            if not rel:
                continue
            dest = work_dir / rel
            dest.parent.mkdir(parents=True, exist_ok=True)
            content = extra.get("content") or b""
            if isinstance(content, bytes):
                dest.write_bytes(content)
            else:
                dest.write_text(str(content), encoding="utf-8")
            asset_entries.append(
                {
                    "filename": rel,
                    "mimeType": (
                        "application/pdf"
                        if dest.suffix.lower() == ".pdf"
                        else "application/octet-stream"
                    ),
                    "size": dest.stat().st_size,
                }
            )

        tex_path.write_text(tex_content, encoding="utf-8")
        asset_entries.append(
            {
                "filename": tex_path.name,
                "mimeType": "application/x-tex; charset=utf-8",
                "size": tex_path.stat().st_size,
            }
        )

        for current_pass in range(1, passes + 1):
            pass_started = time.perf_counter()
            cmd = [
                "pdflatex",
                "-interaction=nonstopmode",
                "-halt-on-error",
                "-file-line-error",
            ]
            # In the two-pass pipeline the first pass prepares auxiliary references.
            if passes == 2 and current_pass == 1:
                cmd.append("-draftmode")
            cmd.append(tex_path.name)
            run = subprocess.run(
                cmd,
                cwd=work_dir,
                capture_output=True,
                text=True,
                timeout=timeout,
                check=False,
            )
            output = f"{run.stdout or ''}\n{run.stderr or ''}"
            logs.append(f"--- pdflatex pass {current_pass} ---\n{output}")
            profile[f"pdflatex_pass_{current_pass}_ms"] = (time.perf_counter() - pass_started) * 1000
            if run.returncode != 0:
                raise LatexCompilationError(
                    "compilation_failed",
                    f"pdflatex failed on pass {current_pass} with status {run.returncode}.",
                    logs="\n".join(logs),
                    asset_manifest=build_asset_manifest(asset_entries),
                    work_dir=str(work_dir) if preserve_workdir_on_error else None,
                )

        if not pdf_path.exists():
            raise LatexCompilationError(
                "output_missing",
                f"Expected PDF output was not generated at {pdf_path}.",
                logs="\n".join(logs),
                asset_manifest=build_asset_manifest(asset_entries),
                work_dir=str(work_dir) if preserve_workdir_on_error else None,
            )

        read_started = time.perf_counter()
        pdf_buffer = pdf_path.read_bytes()
        profile["read_pdf_ms"] = (time.perf_counter() - read_started) * 1000
        return {
            "pdfBuffer": pdf_buffer,
            "logs": "\n".join(logs),
            "profile": profile,
            "workDir": str(work_dir),
            "assetManifest": build_asset_manifest(asset_entries),
        }
    finally:
        cleanup_started = time.perf_counter()
        if cleanup and not preserve_workdir_on_error:
            shutil.rmtree(work_dir, ignore_errors=True)
        profile["cleanup_ms"] = (time.perf_counter() - cleanup_started) * 1000
